### PR TITLE
Elektra High Level API

### PR DESCRIPTION
**Pull request by domhof**
# Purpose

First draft of Elektra High Level API.

# Checklist

Please only check relevant points.
For docu fixes, spell checking and similar nothing
needs to be checked.

- [ ] commit messages are fine (with references to issues)
- [ ] I ran all tests and everything went fine
- [ ] I added unit tests
- [ ] affected documentation is fixed
- [ ] I added code comments, logging, and assertions
- [ ] meta data is updated (e.g. README.md of plugins)

@markus2330 please review my pull request

### CODE REVIEW

**Comment by markus2330** [comment]
Thank you!

Can you put elektra.c into a separate library? It won't compile (and does not make sense) to put it into the core.

@tom-wa Can you do a review, too?

File: src/include/elektra.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ */
+
+#ifndef ELEKTRA_H
+#define ELEKTRA_H
+
+#include "kdb.h"
+#include "kdbhelper.h"
+#include "kdbtypes.h"
+
+struct _ElektraError
```
Comment (markus2330) [inline src/include/elektra.h]: Structs should be hidden!

File: src/include/elektra.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+{
+    int errorNr;
```
Comment (markus2330) [inline src/include/elektra.h]: Why is this needed? Do you want to change elektraHasError to return a kdb_long_t?

File: src/include/elektra.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    const char * msg;
```
Comment (markus2330) [inline src/include/elektra.h]: Shouldn't this be an array if you can hold errorNr errors?

File: src/libs/elektra/elektra.c [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+const char * elektraErrorMessage (const Elektra * elektra)
+{
+    return elektra->error->msg;
+}
+
+void elektraClearError (Elektra * elektra)
+{
+    elektraFree (elektra->error);
+    elektra->error = NULL;
```
Comment (markus2330) [inline src/libs/elektra/elektra.c]: Nice and simple!

File: src/libs/elektra/elektra.c [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+{
+    Key * const nameKey = keyDup (elektra->parentKey);
+    keyAddName (nameKey, name);
+
+    Key * const resultKey = ksLookup (elektra->config, nameKey, 0);
+    const char * string = keyString (resultKey);
+
+    keyDel (nameKey);
+
+    const kdb_long_long_t value = ELEKTRA_LONG_LONG_S (string, NULL, 10);
```
Comment (markus2330) [inline src/libs/elektra/elektra.c]: use keyGetLongLong here?

File: src/libs/elektra/elektra.c [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+ * @brief Elektra High Level API.
+ *
+ * @copyright BSD License (see doc/LICENSE.md or http://www.libelektra.org)
+ */
+
+#include "elektra.h"
+#include <stdlib.h>
+#include <memory.h>
+
+Elektra * elektraOpen (const char * application)
```
Comment (markus2330) [inline src/libs/elektra/elektra.c]: Please add error handling here!

It is important to understand which guarantees we can give to libraries.

File: src/include/elektra.h (hunk 2 above)
Comment (domhof) [inline src/include/elektra.h]: Yes, we could do something like that (maybe with another function). The idea is that the developer can identify an error-type without parsing the string. Maybe it is not needed, I would say we will see which error-types we will have and can then decide upon that if we keep it or remove it.

File: src/include/elektra.h (hunk 3 above)
Comment (domhof) [inline src/include/elektra.h]: errorNr just identifies the error-type. msg is a readable string that describes it.

File: src/libs/elektra/elektra.c (hunk 5 above)
Comment (domhof) [inline src/libs/elektra/elektra.c]: As far as I know we only have getString and getValue in the layer below. (we decided to remove all typed getters)

File: src/include/elektra.h (hunk 2 above)
Comment (markus2330) [inline src/include/elektra.h]: > The idea is that the developer can identify an error-type without parsing the string.

Yes, sounds useful. For the code generator option we could even provide an enum.

> I would say we will see which error-types we will have

They are already defined in `src/error/specification`.

File: src/libs/highlevel/elektra.c [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+/**
+ * @param elektra The elektra instance initialized with the parent key.
+ * @param name The keyname to look up. The keyname is appended to the parent key.
+ */
+kdb_boolean_t elektraGetBoolean (Elektra * elektra, const char * name)
+{
+    READ_KEY
+
+    kdb_boolean_t value = 0;
+    if (!strcmp(string, "true") || !strcmp(string, "1") || !strcmp(string, "on")) {
```
Comment (markus2330) [inline src/libs/highlevel/elektra.c]: Here you introduce the same issue as discussed with hex values in #1388

We already have defined booleans to be `0` or `1`.

Furthermore, you need to check here if `check/type` isn't something else other than `boolean`.

File: src/include/elektra.h [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+#ifndef ELEKTRA_H
+#define ELEKTRA_H
+
+#include "kdbtypes.h"
+
+typedef struct _ElektraError ElektraError;
+typedef struct _Elektra Elektra;
+
+Elektra * elektraOpen (const char * application);
```
Comment (markus2330) [inline src/include/elektra.h]: error parameter missing

File: src/include/elektra.h [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+void elektraClose (Elektra * elektra);
+
+kdb_boolean_t elektraHasError (const Elektra * elektra);
```
Comment (markus2330) [inline src/include/elektra.h]: old API

File: src/include/elektraprivate.h [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    const char * msg;
+};
+
+struct _Elektra
+{
+    KDB * kdb;
+    KeySet * config;
+    Key * parentKey;
+
+    struct _ElektraError * error;
```
Comment (markus2330) [inline src/include/elektraprivate.h]: not needed

File: src/libs/CMakeLists.txt [hunk 11]
```diff
@@ -15,3 +15,5 @@ add_subdirectory (plugin)
 add_subdirectory (elektra)
 
 generate_manpage (elektra-libs FILENAME ${CMAKE_CURRENT_SOURCE_DIR}/README.md SECTION 7)
+
+add_subdirectory (highlevel)
```
Comment (markus2330) [inline src/libs/CMakeLists.txt]: please add newline

File: src/libs/highlevel/elektra.c [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+Key * const nameKey = keyDup (elektra->parentKey); \
+keyAddName (nameKey, name); \
+\
+Key * const resultKey = ksLookup (elektra->config, nameKey, 0); \
+const char * string = keyString (resultKey); \
+\
+keyDel (nameKey); \
+
+#define RELEASE_KEY_AND_RETURN_VALUE \
+keyDel (resultKey); \
```
Comment (markus2330) [inline src/libs/highlevel/elektra.c]: is this macro really needed?

File: src/libs/highlevel/elektra.c [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+
+#include "elektra.h"
+#include <stdlib.h>
+#include <memory.h>
+#include "kdb.h"
+#include "kdbhelper.h"
+#include "elektraprivate.h"
+
+#define READ_KEY \
+Key * const nameKey = keyDup (elektra->parentKey); \
```
Comment (markus2330) [inline src/libs/highlevel/elektra.c]: Why duplicate? What about having a searchKey in `Elektra` and only modify its name?

File: src/libs/highlevel/elektra.c [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+
+/**
+ * @param elektra The elektra instance initialized with the parent key.
+ * @param name The keyname to look up. The keyname is appended to the parent key.
+ */
+const char * elektraGetString (Elektra * elektra, const char * name)
+{
+    READ_KEY
+
+    char * value = elektraMalloc (keyGetValueSize(resultKey));
```
Comment (markus2330) [inline src/libs/highlevel/elektra.c]: Do not return allocated memory to the user.

File: src/tools/gen/support/c.py [hunk 15]
```diff
@@ -13,13 +13,13 @@ def funcname(self, key):
 			raise Exception("invalid keyname " + key)
 
 	def getfuncname(self, key):
-		return "get_"+self.funcname(key)
+		return "elektraGet"+self.funcname(key)
```
Comment (markus2330) [inline src/tools/gen/support/c.py]: This change would need a fix in the tests. It might be better to have a new py file which inherits from `c.py`.

Please cut off basename here, or fix specification of your user study example to not contain `Sw`.

**Comment by markus2330** [comment]
Thanks for working on it!

Seems you need to rebase (conflicing file .gitignore)

Please remove "work in progress" when I should do a review.

File: tests/ctest/test_elektra.c [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    int result = kdbSet (handle, config, parentKey);
+    printf ("SET: result is %d\n", result);
+    if (result == -1) {
+        Key * problemKey = ksCurrent (config);
+        printf ("SET: problemKey is %s\n", keyName (problemKey));
+    }
+
+    // Close
+    kdbClose (handle, parentKey);
+    keyDel (parentKey);
```
Comment (markus2330) [inline tests/ctest/test_elektra.c]: not everything cleaned up, e.g. config

File: tests/ctest/test_elektra.c [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+    printf ("CHECK: value is %s\n", keyString(checkKey));
+
+    // Close
+    kdbClose (handle, parentKey);
+    keyDel (parentKey);
+}
+
+static void test_development ()
+{
+    const char * parentKey = "user/sw/elektra/kdb/#0/current";
```
Comment (markus2330) [inline tests/ctest/test_elektra.c]: Only write in `/test` for test cases, e.g. `user/test/...`

Comment (markus2330) [inline tests/ctest/test_elektra.c]: Btw. the test cases redirect `user` to `/tmp`, so you cannot use `kdb get` to check the key afterwards.

File: tests/ctest/test_elektra.c [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+#include <elektra.h>
+
+static void set (const char * parentKeyName, const char * name, const char * value) {
+    // Open
+    KeySet * config = ksNew (0, KS_END);
+    Key * parentKey = keyNew (parentKeyName, KEY_END);
+    KDB * handle = kdbOpen (parentKey);
+    kdbGet (handle, config, parentKey);
+
+    printf ("Error occurred: %s\n", keyString (keyGetMeta (parentKey, "error/description")));
```
Comment (domhof) [inline tests/ctest/test_elektra.c]: lldb tells me that the handle is NULL ;-)

File: tests/ctest/test_elektra.c [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+
+    // Set
+    Key *key = keyNew(parentKeyName, KEY_END);
+    keyAddName(key, name);
+    keySetString(key, value);
+    ksAppendKey(config, key);
+
+    // Check
+    Key *checkKey = ksLookupByName (config, name, 0);
```
Comment (domhof) [inline tests/ctest/test_elektra.c]: This file was never meant to be added to git, I tried different things until I found out that the KDB could not be loaded in the first place. I'm aware of the inconsistencies, please ignore them, the code will get removed anyway after we have found out why the resolver cannot be loaded (or about the real cause).

File: tests/ctest/test_elektra.c (hunk 16 above)
Comment (domhof) [inline tests/ctest/test_elektra.c]: see comment above

File: tests/ctest/test_elektra.c [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+
+static void check (const char * parentKeyName, const char * name) {
+    // Open
+    KeySet * config = ksNew (0, KS_END);
+    Key * parentKey = keyNew (parentKeyName, KEY_END);
+    KDB * handle = kdbOpen (parentKey);
+    kdbGet (handle, config, parentKey);
+
+    // Check
+    Key *checkKey = ksLookupByName (config, name, 0);
```
Comment (domhof) [inline tests/ctest/test_elektra.c]: see comment above

File: tests/ctest/test_elektra.c (hunk 17 above)
Comment (domhof) [inline tests/ctest/test_elektra.c]: Thank you! I didn't know about user being redirected to tmp. Is there a documentation for writing tests etc.?

File: tests/ctest/test_elektra.c [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+    const char * name = "test";
+    const char * value = "1";
+
+    set (parentKey, name, value);
+    check (parentKey, name);
+}
+
+int main (int argc, char ** argv)
+{
+	printf ("KEY      TESTS\n");
```
Comment (domhof) [inline tests/ctest/test_elektra.c]: See comment above. I initially copied test_key.c. This is just a leftover.

File: tests/ctest/test_elektra.c (hunk 18 above)
Comment (markus2330) [inline tests/ctest/test_elektra.c]: Okay, something seems to be very broken, then.

But in this case the error 37 with the description `handle or ks null pointer` should be set?

File: tests/ctest/test_elektra.c (hunk 19 above)
Comment (markus2330) [inline tests/ctest/test_elektra.c]: Okay, for me it worked except of this.

File: tests/ctest/test_elektra.c (hunk 17 above)
Comment (markus2330) [inline tests/ctest/test_elektra.c]: Only `doc/TESTING.md`. For non-ABI tests it is okay to use gtest, which is much more comfortable and better documented. (And according to @fberlakovich it has some CLion support.)

We can also consider to write ABI tests with gtest. Until now nobody checked what would happen in the case of gtest upgrades. Maybe it will work without issues anyway. (Because gtest is compiled in after all.)

We can also consider a gtest upgrade, we still have 1.7.0 but 1.8.0 was already released.

Comment (markus2330) [inline tests/ctest/test_elektra.c]: Improvements of `doc/TESTING.md` and how we use gtest is of course very welcomed!

File: src/error/exporterrors.cpp [hunk 22]
```diff
@@ -113,7 +113,29 @@ ostream & operator<< (ostream & os, parse_t & p)
 		os << p[i]["macro"] << " " << i << endl;
 	}
 
+    os << endl << endl;
+
+	/* Temporarily disable enum generation.
```
Comment (markus2330) [inline src/error/exporterrors.cpp]: Please revert that change, there is little use of passing all enums to the user.

File: src/include/kdbprivate.h [hunk 23]
```diff
@@ -24,6 +24,32 @@
 
 #include <limits.h>
 
+#define KDB_STRING_TO_BOOLEAN(string) (kdb_boolean_t) !strcmp (string, "1")
```
Comment (markus2330) [inline src/include/kdbprivate.h]: Maybe better in kdbmacros.h? Is this useful without any error handling? Prefer to prefix with ELEKTRA_

File: src/include/elektra_error_private.h [hunk 24]
```diff
@@ -0,0 +1,26 @@
+/**
```
Comment (markus2330) [inline src/include/elektra_error_private.h]: One private header should be enough?

File: src/include/elektra_error_codes.h [hunk 25]
```diff
@@ -0,0 +1,58 @@
+/**
```
Comment (markus2330) [inline src/include/elektra_error_codes.h]: So many errors are not needed.

File: src/include/elektra_error.h [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+typedef enum {
+    ELEKTRA_ERROR_SEVERITY_FATAL = 0,
+    ELEKTRA_ERROR_SEVERITY_ERROR,
+    ELEKTRA_ERROR_SEVERITY_WARNING
+} ElektraErrorSeverity;
+
+typedef const char * ElektraErrorGroup;
+typedef const char * ElektraErrorModule;
+
+ElektraErrorCode elektraErrorCode (ElektraError * error);
```
Comment (markus2330) [inline src/include/elektra_error.h]: rename to elektraErrorGetCode?

File: src/include/elektra.h [hunk 27]
```diff
@@ -0,0 +1,393 @@
...
+#include "kdb.h"
+#include "kdbtypes.h"
+#include "elektra_error.h"
+
+typedef struct _Elektra Elektra;
+
+Elektra * elektraOpen (const char * application, ElektraError ** error);
+void elektraClose (Elektra * elektra);
+
+/**
```
Comment (markus2330) [inline src/include/elektra.h]: docu please in source files. Add more docu, use copyDoc for same content.

File: doc/decisions/high_level_api.md [hunk 28]
```diff
@@ -75,13 +75,18 @@ sure to use pass a `KeySet`, that has your whole specification included
...
+ElektraErrorGroup elektraErrorGroup (ElektraError * error);
+ElektraErrorModule elektraErrorModule (ElektraError * error);
+
+void elektraErrorFree (ElektraError * error);
+
 ```
 
-`elektraErrorAbort` tells you if you need to quit your application because of severe
-issues that are permanent. Otherwise, developers can just print the message and
+`elektraErrorSeverity` tells you if you need to quit your application because of severe
```
Comment (markus2330) [inline doc/decisions/high_level_api.md]: Also add this (or move this) parts to the official docu (tutorial)

File: .gitignore [hunk 29]
```diff
@@ -32,3 +32,7 @@ nbproject/
 #ignore macOS folder attribute files
 .DS_Store
 .directory
+
+#ignore CLion project files
+.idea/
+cmake-build-debug/
```
Comment (markus2330) [inline .gitignore]: also add cmake-build-release/?

File: src/error/specification [hunk 30]
```diff
@@ -24,6 +24,7 @@ module:dl
 severity:warning
 class:FailedLoadModuleError
 extends:ModuleError
+enum:DLOPEN_FAILED
```
Comment (markus2330) [inline src/error/specification]: Great to see all these enum values!

File: src/error/specification [hunk 31]
```diff
@@ -436,650 +506,746 @@ module:dl
...
 module:simpleini
+enum:INVALID_LINE_FORMAT_PATTERN
 
 number:174
 description:general xml error
 severity:error
 ingroup:plugin
 module:xerces
-
-number:175
```
Comment (markus2330) [inline src/error/specification]: Please rebase, you remove something by accident here.

File: src/libs/highlevel/elektra.c [hunk 32]
```diff
@@ -0,0 +1,473 @@
...
+    {
+        *error = elektraErrorCreateFromKey(parentKey);
+        return NULL;
+    }
+
+    Elektra * const elektra = elektraCalloc (sizeof (struct _Elektra));
+    elektra->kdb = kdb;
+    elektra->parentKey = parentKey;
+    elektra->config = config;
+    elektra->lookupKey = keyNew(NULL);
```
Comment (markus2330) [inline src/libs/highlevel/elektra.c]: Please use KEY_END

File: src/libs/highlevel/elektra.c [hunk 33]
```diff
@@ -0,0 +1,473 @@
...
+
+
+Elektra * elektraOpen (const char * application, ElektraError ** error)
+{
+    Key * const parentKey = keyNew (application, KEY_END);
+    KDB * const kdb = kdbOpen (parentKey);
+
+    if (kdb == NULL)
+    {
+        *error = elektraErrorCreateFromKey(parentKey);
```
Comment (markus2330) [inline src/libs/highlevel/elektra.c]: error fatal?

File: src/libs/highlevel/elektra.c [hunk 34]
```diff
@@ -0,0 +1,473 @@
...
+
+        keySetString (key, value);
+
+        ret = kdbSet (elektra->kdb, elektra->config, elektra->parentKey);
+        if (ret == -1)
+        {
+            Key * problemKey = ksCurrent (elektra->config);
+            if (problemKey != NULL)
+            {
+//                ELEKTRA_LOG_DEBUG("problemKey: %s\n", keyName (problemKey));
```
Comment (markus2330) [inline src/libs/highlevel/elektra.c]: Why comment out?

File: src/libs/highlevel/elektra.c [hunk 35]
```diff
@@ -0,0 +1,473 @@
...
+            }
+
+            kdbGet (elektra->kdb, elektra->config, elektra->parentKey);
+        }
+    } while (ret == -1);
```
Comment (markus2330) [inline src/libs/highlevel/elektra.c]: Might not terminate.

File: src/libs/highlevel/elektra.c [hunk 36]
```diff
@@ -0,0 +1,473 @@
...
+        ret = kdbSet (elektra->kdb, elektra->config, elektra->parentKey);
+        if (ret == -1)
+        {
+            Key * problemKey = ksCurrent (elektra->config);
+            if (problemKey != NULL)
+            {
+//                ELEKTRA_LOG_DEBUG("problemKey: %s\n", keyName (problemKey));
+            }
+
+            kdbGet (elektra->kdb, elektra->config, elektra->parentKey);
```
Comment (markus2330) [inline src/libs/highlevel/elektra.c]: Unnecessary kdbGet if kdbSet was successful.

Please take a look at examples/kdbset.c or qt-gui for how to do this correctly. You also need to do a 3-way merge considering who changed the value (for the one key to set).

File: src/libs/highlevel/elektra.c [hunk 37]
```diff
@@ -0,0 +1,473 @@
...
+
+        keySetString (key, value);
+
+        ret = kdbSet (elektra->kdb, elektra->config, elektra->parentKey);
+        if (ret == -1)
+        {
+            Key * problemKey = ksCurrent (elektra->config);
+            if (problemKey != NULL)
+            {
+//                ELEKTRA_LOG_DEBUG("problemKey: %s\n", keyName (problemKey));
```
Comment (markus2330) [inline src/libs/highlevel/elektra.c]: Why comment out? Seems to be duplication of code here?

File: src/libs/highlevel/README.md [hunk 38]
```diff
@@ -0,0 +1,154 @@
...
+{
+  ElektraError * error = NULL;
+  Elektra * elektra = elektraOpen ("/sw/yourApplication", &error);
+
+  if (error != NULL) 
+  {
+    printf ("Sorry, there seems to be an error with your Elektra setup: %s\n", elektraErrorDescription (error));
+    printf ("Will exit now...\n");
+    exit (EXIT_FAILURE);
+    elektraErrorFree (error);
```
Comment (domhof) [inline src/libs/highlevel/README.md]: An error occured and the handle is NULL so elektraClose is not necessary. elektraErrorFree really makes no sense here, swapped the two lines.

File: src/libs/highlevel/README.md [hunk 39]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  const char * welcomeMessage = elektraGetString (elektra, "welcomeMessage");
```
Comment (domhof) [inline src/libs/highlevel/README.md]: Changed it to "message"

File: src/libs/highlevel/README.md [hunk 40]
```diff
@@ -0,0 +1,154 @@
...
+
+## Example
+```c
+#include <stdio.h>
+#include "elektra.h"
+
+int main(int argc, char ** argv)
+{
+  ElektraError * error = NULL;
+  Elektra * elektra = elektraOpen ("/sw/yourApplication", &error);
```
Comment (domhof) [inline src/libs/highlevel/README.md]: Changed it to "/sw/org/myapp/#0/current".

File: src/libs/highlevel/README.md [hunk 41]
```diff
@@ -0,0 +1,154 @@
...
+### Struct `ElektraError`
+The library is designed to hide as many problems a developer can encounter when usign KDB directly. However it is not possible to hide all those issues. As with every library, things can go wrong and there needs to be a way react to errors once they have occurred at runtime. Therefore the High-Level API introduces a simplified struct called `ElektraError`, which encapsulates all information neccessary for the developer to handle runtime-errors appropriately in the application.
+
+Functions that can produce errors accept an ElektraError pointer as parameter, for example:
+
+```c
+Elektra * elektraOpen (const char * application, ElektraError ** error);
+```
+
+In most cases you'll want to set the error variable to NULL before passing it to the function. You can do this either by declaring and initializing a new variable with `ElektraError * error = NULL` or by reusing an already existing error variable by resetting it with `elektraErrorFree (error)`.
```
Comment (domhof) [inline src/libs/highlevel/README.md]: Thank you for pointing this out! In fact it was necessary. I changed the name to elektraErrorReset and the function now in addition resets the passed error variable to NULL. Seems to me more intuitive this way.

File: src/libs/highlevel/README.md [hunk 42]
```diff
@@ -0,0 +1,154 @@
...
+First you have to add `elektra-highlevel` to the linked libraries of your application. To be able to use it in your source file, just include the main header with `#include "elektra.h"` at the top of your file.
+
+## Core Concepts
+
+### Struct `Elektra`
+`Elektra` is the handle you use to access the underlying KDB. All key-value read and write operations expect this handle to be passed as in as a parameter. To create the handle, you can simply write:
+
+```c
+ElektraError * error = NULL;
+Elektra * elektra = elektraOpen ("/sw/yourApplication", &error);
```
Comment (domhof) [inline src/libs/highlevel/README.md]: I think the user gets the point. Otherwise he would ask himself what the NULL parameter could be.

File: src/bindings/cpp/benchmarks/benchmark_thread.cpp [hunk 43]
```diff
@@ -67,8 +67,9 @@ class Environment : public ThreadBoolean
 	: ThreadBoolean (ks, context_, Key ("/test/%layer1%", KEY_CASCADING_NAME, KEY_META, "default", "1", KEY_END)),
 	  nested (ks, context_), person (ks, context_),
 	  profile (ks, context_, Key ("/%layer1%/profile", KEY_CASCADING_NAME, KEY_META, "default", "default", KEY_END)),
-	  bm (ks, context_, Key ("/%layer1%/%layer2%/%layer3%/%layer4%/%layer5%/%layer6%/%layer7%/%layer8%/%layer9%/", KEY_CASCADING_NAME,
```
Comment (markus2330) [inline src/bindings/cpp/benchmarks/benchmark_thread.cpp]: Did you use clang-format-3.8? Maybe @sanssecours can give you a tip how to get it.

If we would use different versions of `clang-format` we would keep reverting our changes.

Comment (sanssecours) [inline src/bindings/cpp/benchmarks/benchmark_thread.cpp]: > Did you use clang-format-3.8? Maybe @sanssecours can give you a tip how to get it (on Mac).

You can install `clang-3.8` via `brew install clang-format@3.8`.

> If we would use different versions of `clang-format` we would keep reverting our changes.

We could also just use the current version (`5.0.0`) of the software.

Comment (markus2330) [inline src/bindings/cpp/benchmarks/benchmark_thread.cpp]: Thanks for giving the tip!

> We could also just use the current version (5.0.0) of the software.

Installing clang on all build agents take some time, so I would prefer to wait for @BernhardDenner's build server case study. Furthermore, it needs to be coordinated with all developers. We can discuss it in the next meeting, but for now @domhof please use 3.8.

Comment (domhof) [inline src/bindings/cpp/benchmarks/benchmark_thread.cpp]: Thanks, I used the current version (5.0.0). Reverted back to 3.8 for now.

**Comment by markus2330** [comment]
Btw. before merging please rebase. And it would be great if you can put the module you are working on as short message before the commit messages. (kdbhl or similar)

And as long term perspective we should get rid of the python dep for the simple parts of the code generator (generating the key names and default config).

**Comment by markus2330** [comment]
What is the status of this PR? Still seems to have a lot of wrong reformatting? Please check the PR at github.

**Comment by markus2330** [comment]
Btw. last time we discussed you will start a new PR. What is the status? Please also make sure to work in all feedback you got so far and self-review what you have done.
